File: javascript/22.generate-parentheses.js
/*
 * @lc app=leetcode id=22 lang=javascript
 *
 * [22] Generate Parentheses
 */

// @lc code=start
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
  const parentheses = [];
  function backtrackInsert(l, r, s) {
    if (l === n && r === n) return parentheses.push(s);
    if (l < n) {
      backtrackInsert(l + 1, r, s + "(");
    }
    if (r < l && r < n) {
      backtrackInsert(l, r + 1, s + ")");
    }
  }
  backtrackInsert(0, 0, "");
  return parentheses;
};
// @lc code=end

File: rust/22.generate-parentheses.rs
/*
 * @lc app=leetcode id=22 lang=rust
 *
 * [22] Generate Parentheses
 */

// @lc code=start
impl Solution {
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let mut parentheses: Vec<String> = vec![];
        fn backtrack_insert(
            l: i32,
            r: i32,
            n: i32,
            parenthesis: String,
            parentheses: &mut Vec<String>,
        ) {
            if l == n && r == n {
                parentheses.push(parenthesis);
                return;
            }
            if l < n {
                backtrack_insert(l + 1, r, n, format!("{parenthesis}("), parentheses);
            }
            if r < l && r < n {
                backtrack_insert(l, r + 1, n, format!("{parenthesis})"), parentheses);
            }
        }
        backtrack_insert(0, 0, n, String::new(), &mut parentheses);
        parentheses
    }
}
// @lc code=end
